File: app/src/main/java/nl/rsdt/japp/jotial/maps/wrapper/osm/OsmCircle.java
package nl.rsdt.japp.jotial.maps.wrapper.osm;

import com.google.android.gms.maps.model.CircleOptions;
import com.google.android.gms.maps.model.LatLng;

import org.osmdroid.util.GeoPoint;
import org.osmdroid.views.MapView;
import org.osmdroid.views.overlay.Polygon;

import java.util.ArrayList;

import nl.rsdt.japp.jotial.maps.wrapper.ICircle;

/**
 * Created by mattijn on 08/08/17.
 */

public class OsmCircle implements ICircle {
    private final MapView osmMap;
    private final Polygon osmCircle;
    private LatLng centre;

    public OsmCircle(CircleOptions circleOptions, MapView osmMap) {
        this.osmMap = osmMap;
        osmCircle = new Polygon();

        osmCircle.setStrokeWidth(circleOptions.getStrokeWidth());
        osmCircle.setFillColor(circleOptions.getFillColor());
        osmCircle.setStrokeColor(circleOptions.getStrokeColor());

        final double radius = circleOptions.getRadius();
        this.centre = circleOptions.getCenter();
        this.setRadius((float) radius);
        osmMap.getOverlays().add(osmCircle);
    }

    public void remove() {
        osmMap.getOverlays().remove(osmCircle) ;
    }

    public void setRadius(float radius) {
        int nrOfPoints = 360;
        ArrayList<GeoPoint> circlePoints = new ArrayList<GeoPoint>();
        for (float f = 0; f < nrOfPoints; f ++){
            circlePoints.add(new GeoPoint(centre.latitude , centre.longitude ).destinationPoint(radius, f));
        }
        osmCircle.setPoints(circlePoints);
    }

    @Override
    public void setVisible(boolean visible) {
        osmCircle.setVisible(visible);
    }

    @Override
    public int getFillColor() {
        return osmCircle.getFillColor();
    }

    @Override
    public void setFillColor(int color) {
        osmCircle.setFillColor(color);
    }
}

File: app/src/main/java/nl/rsdt/japp/jotial/maps/wrapper/osm/OsmCircle.kt
package nl.rsdt.japp.jotial.maps.wrapper.osm

import com.google.android.gms.maps.model.CircleOptions
import com.google.android.gms.maps.model.LatLng
import nl.rsdt.japp.jotial.maps.wrapper.ICircle
import org.osmdroid.util.GeoPoint
import org.osmdroid.views.MapView
import org.osmdroid.views.overlay.Polygon
import java.util.*

/**
 * Created by mattijn on 08/08/17.
 */

class OsmCircle(circleOptions: CircleOptions, private val osmMap: MapView) : ICircle {
    private val osmCircle: Polygon
    private val centre: LatLng

    override var fillColor: Int
        get() = osmCircle.fillColor
        set(color) {
            osmCircle.fillColor = color
        }

    init {
        osmCircle = Polygon()

        osmCircle.strokeWidth = circleOptions.strokeWidth
        osmCircle.fillColor = circleOptions.fillColor
        osmCircle.strokeColor = circleOptions.strokeColor

        val radius = circleOptions.radius
        this.centre = circleOptions.center
        this.setRadius(radius.toFloat())
        osmMap.overlays.add(osmCircle)
    }

    override fun remove() {
        osmMap.overlays.remove(osmCircle)
    }

    override fun setRadius(radius: Float) {
        val nrOfPoints = 360
        val circlePoints = ArrayList<GeoPoint>()
        for (f in 0 until nrOfPoints) {
            circlePoints.add(GeoPoint(centre.latitude, centre.longitude).destinationPoint(radius.toDouble(), f.toFloat()))
        }
        osmCircle.points = circlePoints
    }

    override fun setVisible(visible: Boolean) {
        osmCircle.isVisible = visible
    }
}
